### DISCUSSION TITLE
Getting started tutorial with basic networking

### DISCUSSION THREAD

**Discussion by maxweber** [cloud-hypervisor/cloud-hypervisor#6269] [Q&A]
Your [getting started tutorial](https://github.com/cloud-hypervisor/cloud-hypervisor?tab=readme-ov-file#2-getting-started) is good. Despite the fact that I didn't had much experiences with microvms I managed to get the example up and running (snapshotting also worked). However, I was completely lost when I tried to connect the VM to the network and the internet. I would be very thankful for additional steps in the getting started tutorial that describes how to do a basic network setup.

I was able to get the [Firecracker getting started example](https://github.com/firecracker-microvm/firecracker/blob/main/docs/getting-started.md#starting-firecracker) running with networking. I also read a lot about tap devices and Linux networking in general. But since days I'm not able to get the cloud-hypervisor VM connected to the network.

Thanks a lot for creating this project. I would be very glad about hints, how to get the networking correct.

Best regards,

Max

**Comment by lebron8dong** [comment]
Hello, have you solved the network connection problem ?

**Reply by maxweber** [reply]
@lebron8dong Hi, regrettably, I did not solve the network connection problem.

**Reply by rbradford** [reply]
Is this not sufficient - https://www.cloudhypervisor.org/docs/prologue/quick-start/ ?

**Reply by maxweber** [reply]
@rbradford I tried to follow this guide multiple times and read articles like [this one](https://blog.cloudflare.com/virtual-networking-101-understanding-tap) to get a better understanding. However, my knowledge of networking in Linux seems insufficient to configure a "hello world" example in combination with cloud-hypervisor. Therefore I would be very thankful for any help and hints on how do to this.

**Reply by vkhamsao** [reply]
For just a hello world example this would suffice, definitely wouldn't recommend for any long term use cases.  This requires nftables and iproute2  on the hypervisor, which should be standard for most distros.

## Hypervisor
```bash
tap_if=vmtap0
bridge=vmbr0

sudo mkdir -p /etc/nftables

sudo tee /etc/nftables/masquerade.nft >/dev/null <<EOF
#!/usr/sbin/nft -f

flush ruleset

table ip nat {
    chain postrouting {
        type nat hook postrouting priority 0;
        iifname "$bridge" masquerade
    }
}
EOF

sudo nft -f /etc/nftables/masquerade.nft
sudo ip link add "$bridge" type bridge
sudo ip address add 172.30.0.1/23 dev "$bridge"
sudo ip link set "$bridge" type up
sudo ip link set "$tap_if" master "$bridge"
sudo sysctl -w net.ipv4.conf.all.forwarding=1
sudo sysctl -w net.ipv6.conf.all.forwarding=1
```

## Guest
```bash
netdev=ens4  # replace as neeeded
sudo ip link set "$netdev" up
sudo ip address add 172.30.0.2/23 dev "$netdev"
sudo ip route add default via 172.30.0.1 dev "$netdev"
ping 1.1.1.1
```

**Reply by maxweber** [reply]
@vkhamsao awesome, thank you very much :-)
